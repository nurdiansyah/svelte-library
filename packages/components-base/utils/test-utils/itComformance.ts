import type { RenderResult } from "@testing-library/svelte";
import type { SvelteComponentDev } from "svelte/internal";
import { cleanup, render } from "@testing-library/svelte";

export type BaseProps = {
  "data-testid"?: string;
};

export type Options<P = Record<string, any>> = {
  props: P;
  container?: HTMLElement;
  parentProps?: Record<string, any>;
  after?: () => void;
  only?: string[];
  skip?: string[];
  target?: HTMLElement;
  className?: string;
};

function testMergeClassName(_render: () => RenderResult, { props, className }: Options) {
  it("cek merge class", () => {
    let el: undefined | Element;
    const { getByTestId, container } = _render();
    if (props["data-testid"]) {
      el = getByTestId(props["data-testid"]);
    } else {
      el = container?.firstElementChild || undefined;
    }
    className && expect(el).toHaveClass(className);
    props["class"] && expect(el).toHaveClass(props["class"]);
  });
}

function testRendererComponent(_render: () => RenderResult, { props }: Options) {
  it("cek render komponen tanpa error dan snapshot", () => {
    let el: undefined | HTMLElement;
    const { getByTestId, container } = _render();
    if (props["data-testid"]) {
      expect(() => (el = getByTestId(props["data-testid"]))).not.toThrow();
      expect(el).toBeInTheDocument();
    }
    expect(container).toMatchSnapshot();
  });
}

const fullSuite: Record<string, (_render: () => RenderResult, options: Options) => void> = {
  renderComponent: testRendererComponent,
  mergeClassName: testMergeClassName
};

export function itConformance<P = Record<string, any>>(component: typeof SvelteComponentDev, options: Options<P>) {
  const { after: runAfterHook = () => {}, only = Object.keys(fullSuite), skip = [], props = {}, container } = options;
  const renderComponent = (_props?: P): RenderResult => {
    const __props = { ...props, ..._props };

    return render(component, __props, { container });
  };
  afterEach(() => {
    cleanup();
    runAfterHook && runAfterHook();
  });

  Object.keys(fullSuite)
    .filter((testKey) => only.indexOf(testKey) !== -1 && skip.indexOf(testKey) === -1)
    .forEach((testKey) => {
      const test = fullSuite[testKey];
      test(renderComponent, options);
    });
  return renderComponent;
}
