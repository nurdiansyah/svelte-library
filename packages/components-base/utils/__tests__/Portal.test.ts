import { render } from "@testing-library/svelte";
import PortalMock from "./features/PortalMock.svelte";
import { BaseProps, itConformance } from "../test-utils";

type Props = BaseProps & Record<string, any>;

describe("Portal", () => {
  const props: Props = { "data-testid": "portal" };

  itConformance(PortalMock, { props });
  const renderTest = (__props: Props = {}) => render(PortalMock, { props: { ...props, ...__props } });
  describe("props", () => {
    it("disable", () => {
      const { getByTestId, container } = renderTest({ disable: true });
      expect(() => getByTestId("container")).not.toThrow();
      expect(container).toMatchSnapshot();
    });
  });
});
