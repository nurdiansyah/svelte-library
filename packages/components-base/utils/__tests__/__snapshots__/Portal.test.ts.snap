// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal cek render komponen tanpa error dan snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        data-testid="container"
      >
        <div
          data-testid="portal"
        >
          <div
            data-testid="child"
          />
        </div>
      </div>
       
      <div
        style="display: none;"
      />
      
    </div>
  </div>
</body>
`;

exports[`Portal props disable 1`] = `
<body>
  <div>
    <div>
      <div
        data-testid="container"
      />
       
      <div
        data-testid="child"
      />
      
    </div>
  </div>
</body>
`;
